cmd/rofl/deploy.go

```diff
@@ -0,0 +1,297 @@
+			switch {
+			case err == nil:
+			case errors.Is(err, os.ErrNotExist):
+				cobra.CheckErr(fmt.Sprintf("ROFL app bundle '%s' not found. Run `oasis rofl build` first.", orcFilename))
+			default:
+				cobra.CheckErr(fmt.Sprintf("Failed to push ROFL app to OCI repository: %s", err))
+			}
+
+			// Define the deployment.
+			machineDeployment := roflmarket.Deployment{
+				AppID:        appID,
+				ManifestHash: manifestHash,
+				Metadata: map[string]string{
+					"net.oasis.deployment.orc.ref": fmt.Sprintf("%s@%s", deployment.OCIRepository, ociDigest),
+				},
+			}
+
+			switch machine.ID {
+			case "":
+				// When machine is not set, we need to obtain one.
+				fmt.Printf("No pre-existing machine configured, creating a new one...\n")
+
+				if machine.Offer == "" && deployOffer == "" {
+					// Display all offers supported by the provider.
+					showProviderOffers(ctx, npa, conn, *providerAddr)
+					cobra.CheckErr(fmt.Sprintf("Offer not configured for deployment '%s' machine '%s'. Please specify --offer.", deploymentName, deployMachine))
+				}
+				if deployOffer != "" {
+					machine.Offer = deployOffer
+				}
+
+				// Resolve offer.
+				var offers []*roflmarket.Offer
+				offers, err = conn.Runtime(npa.ParaTime).ROFLMarket.Offers(ctx, client.RoundLatest, *providerAddr)
+				if err != nil {
+					cobra.CheckErr(fmt.Sprintf("Failed to query provider: %s", err))
+				}
+				var offer *roflmarket.Offer
+				for _, of := range offers {
+					if of.Metadata[provider.SchedulerMetadataOfferKey] == machine.Offer {
+						offer = of
+						break
+					}
+				}
+				if offer == nil {
+					showProviderOffers(ctx, npa, conn, *providerAddr)
+					cobra.CheckErr(fmt.Sprintf("Offer '%s' not found for provider '%s'.\n", machine.Offer, providerAddr))
+					return
+				}
+
+				fmt.Printf("Taking offer: %s [%s]\n", machine.Offer, offer.ID)
+
+				// Prepare transaction.
+				tx := roflmarket.NewInstanceCreateTx(nil, &roflmarket.InstanceCreate{
+					Provider:   *providerAddr,
+					Offer:      offer.ID,
+					Deployment: &machineDeployment,
+					Term:       term,
+					TermCount:  deployTermCount,
+				})
+
+				acc := common.LoadAccount(cfg, npa.AccountName)
+				var sigTx, meta any
+				sigTx, meta, err = common.SignParaTimeTransaction(ctx, npa, acc, conn, tx, nil)
+				cobra.CheckErr(err)
+
+				var instanceID roflmarket.InstanceID
+				if !common.BroadcastOrExportTransaction(ctx, npa, conn, sigTx, meta, &instanceID) {
+					return
+				}
+
+				rawInstanceID, _ := instanceID.MarshalText()
+				cobra.CheckErr(err)
+				machine.ID = string(rawInstanceID)
+
+				fmt.Printf("Created machine: %s\n", machine.ID)
+			default:
+				// Deploy into existing machine.
+				var instanceID roflmarket.InstanceID
+				if err = instanceID.UnmarshalText([]byte(machine.ID)); err != nil {
+					cobra.CheckErr(fmt.Errorf("malformed machine ID: %w", err))
+				}
+
+				fmt.Printf("Deploying into existing machine: %s\n", machine.ID)
```

Comment:
Let's check, if app ID on the machine matches our app ID (upgrades). And then require `--force` flag, if it doesn't, so people don't deploy ROFL to an existing machine by mistake.

Reply:
Added.